### PR TITLE
Add telemetry logging for unexpected exceptions in Microsoft.VisualSt…

### PR DESCRIPTION
**Pull request by mavasani**
…udio.Editors

**Logging Infrastructure:** We are using the VS Telemetry APIs to log actionable exceptions, this is the recommended approach across DevDiv for Dev15.

**Added API:** Added a utility API "ReportWithoutCrash" that given an exception, logs telemetry for the exception with context (name of the throwing component and additional context message) and returns whether or not the exception is recoverable. API has optional flags to trigger Debug.Fail or treat given exception as always recoverable (catch all filter).

**Code changes in the project to consume this API:** API is always invoked in the exception filter for catch clauses. Cases where this API invocation was added:

1. Debug.Fail within catch clause: We surely care about telemetry for these cases.
2. Existing invocations of Utils.IsUnrecoverableException in catch filters where telemetry might help fix/cleanup code: Used my best judgement on whether telemetry data is helpful, erring on conservative side of more data when I was unsure.
3. [WIP] Silently swallowing System.Exception without re-throwing. It is recommended to catch specific exception kinds for scenarios where we know an operation might throw. However, there are tons of such occurrences across the project, and it is not clear if we going to clean up all of them. I have changed few places to add invocations to ReportWithoutCrash, but fixing over the entire project would need someone familiar with the code to do an audit.

**Related issue:** https://github.com/dotnet/roslyn-project-system/issues/41

### CODE REVIEW

File: src/Microsoft.VisualStudio.Editors/AttributeEditor/PermissionSetService.vb [hunk 1]
```diff
@@ -113,12 +114,12 @@ Namespace Microsoft.VisualStudio.Editors.VBAttributeEditor
...
-                Return SecurityUtilities.ComputeZonePermissionSet( _
-                    strTargetZone, _
-                    projectPermissionSet, _
+                Return SecurityUtilities.ComputeZonePermissionSet(
+                    strTargetZone,
+                    projectPermissionSet,
                     identityList)
 
-            Catch ex As Exception
+            Catch ex As Exception When Common.Utils.ReportWithoutCrash(ex, NameOf(ComputeZonePermissionSet), NameOf(PermissionSetService), considerExceptionAsRecoverable:=True)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/AttributeEditor/PermissionSetService.vb]: Example for Category 3 (Silently swallowing System.Exception without re-throwing)

File: src/Microsoft.VisualStudio.Editors/Common/CodeModelUtils.vb [hunk 2]
```diff
@@ -143,9 +143,7 @@ Namespace Microsoft.VisualStudio.Editors.Common
 
                 'And navigate...
                 TextPoint.Parent.Selection.MoveToPoint(TextPoint)
-            Catch ex As Exception
-                RethrowIfUnrecoverable(ex)
-                Debug.Fail("Navigate to function '" & Func.Name & "' failed: " & ex.ToString)
+            Catch ex As Exception When Common.ReportWithoutCrash(ex, $"Navigate to function '{Func.Name}' failed", NameOf(CodeModelUtils), debugFail:=True)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/CodeModelUtils.vb]: Example for Category 1 (Debug.Fail within catch clause)

File: src/Microsoft.VisualStudio.Editors/Common/Utils.vb [hunk 3]
```diff
@@ -801,8 +816,7 @@ Namespace Microsoft.VisualStudio.Editors.Common
                     Try
                         'Path needs a backslash at the end, or it will be interpreted as a directory + filename
                         InitialDirectory = Path.GetFullPath(AppendBackslash(InitialDirectory))
-                    Catch ex As Exception
-                        Common.RethrowIfUnrecoverable(ex)
+                    Catch ex As Exception When Not IsUnrecoverableException(ex)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: Example for Category 2 (Telemetry not helpful for exception)

File: src/Microsoft.VisualStudio.Editors/DesignerFramework/BaseDesignerLoader.vb [hunk 4]
```diff
@@ -676,8 +675,7 @@ Namespace Microsoft.VisualStudio.Editors.DesignerFramework
                 Dim LostFocusProjectItem As EnvDTE.ProjectItem = Nothing
                 Try
                     LostFocusProjectItem = LostFocus.ProjectItem
-                Catch ex As Exception
-                    Common.RethrowIfUnrecoverable(ex)
+                Catch ex As Exception When Common.ReportWithoutCrash(ex, NameOf(m_WindowEvents_WindowActivated), NameOf(BaseDesignerLoader))
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/DesignerFramework/BaseDesignerLoader.vb]: Example for Category 2 (Not sure if telemetry is helpful here, but still logging to err on conservative side).

**Comment by mavasani** [comment]
Tagging @dotnet/project-system for review.

File: src/Microsoft.VisualStudio.Editors/Common/Utils.vb [hunk 5]
```diff
@@ -314,36 +314,64 @@ Namespace Microsoft.VisualStudio.Editors.Common
...
         ''' <remarks></remarks>
-        Public Function IsUnrecoverable(ByVal ex As Exception, Optional ByVal IgnoreOutOfMemory As Boolean = False) As Boolean
-            If TypeOf ex Is NullReferenceException _
-                OrElse (Not IgnoreOutOfMemory AndAlso TypeOf ex Is OutOfMemoryException) _
-                OrElse TypeOf ex Is StackOverflowException _
-                OrElse TypeOf ex Is System.Threading.ThreadAbortException _
-                OrElse TypeOf ex Is AccessViolationException _
-            Then
-                Return True
+        Public Function ReportWithoutCrash(ByVal ex As Exception,
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: This is the new API to log exceptions.

File: src/Microsoft.VisualStudio.Editors/Common/Utils.vb [hunk 6]
```diff
@@ -314,36 +314,64 @@ Namespace Microsoft.VisualStudio.Editors.Common
...
+            Catch ex As Exception When Not IsUnrecoverableException(ex)
                 Return "[" & ex.GetType.Name & "]"
             End Try
 #Else
             Return ""
 #End If
         End Function
 
+        Public Function IsUnrecoverableException(ByVal ex As Exception, Optional ByVal ignoreOutOfMemory As Boolean = False) As Boolean
+            Return TypeOf ex Is NullReferenceException _
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: Question: Why is NRE considered an unrecoverable exception? Seems very likely that caller can gracefully handle it. Additionally, the definition of the same method in MS.VS.AppDesigner doesn't include NRE (https://github.com/dotnet/roslyn-project-system/blob/master/src/Microsoft.VisualStudio.AppDesigner/Common/Utils.vb#L233)

File: src/Microsoft.VisualStudio.Editors/SettingsDesigner/ConnectionStringUITypeEditor.vb [hunk 7]
```diff
@@ -159,7 +159,7 @@ Namespace Microsoft.VisualStudio.Editors.SettingsDesigner
                     ' Well, we better return the old value...
                     Return oValue
                 End If
-            Catch ex As Exception When Not Common.IsUnrecoverable(ex)
+            Catch ex As Exception When Common.IsUnrecoverableException(ex)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/SettingsDesigner/ConnectionStringUITypeEditor.vb]: Removed "Not" accidentally, will fix.

File: src/Microsoft.VisualStudio.Editors/Common/Utils.vb [hunk 8]
```diff
@@ -314,36 +314,64 @@ Namespace Microsoft.VisualStudio.Editors.Common
...
+                                           Optional ByVal throwingComponentName As String = "general",
+                                           Optional ByVal debugFail As Boolean = False,
+                                           Optional ByVal debugWriteLine As Boolean = False,
+                                           Optional ByVal considerExceptionAsRecoverable As Boolean = False,
+                                           Optional ByVal ignoreOutOfMemory As Boolean = False) As Boolean
+            Debug.Assert(ex IsNot Nothing)
+            Debug.Assert(Not String.IsNullOrEmpty(throwingComponentName))
+            Debug.Assert(Not String.IsNullOrEmpty(exceptionEventDescription))
+
+            TelemetryService.DefaultSession.PostFault(
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: More information about this VSTelemetry API over [here](http://vstelwiki/wiki/Fault_Event)

File: src/Microsoft.VisualStudio.Editors/Common/Utils.vb [hunk 9]
```diff
@@ -519,7 +534,7 @@ Namespace Microsoft.VisualStudio.Editors.Common
         ''' 
         '''       returns the string 
         ''' 
-        '''       "Metafiles (*.wmf, *.emf)|*.wmf;*.emf"
+        '''       "Metafiles (*.wmf, * .emf)|*.wmf;*.emf"
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: is this change intended?

Comment (mavasani) [inline src/Microsoft.VisualStudio.Editors/Common/Utils.vb]: Nope, probably formatter did it while I was editing code above :). I'll revert.

File: src/ProjectSystem.sln [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.25402.0
```
Comment (basoundr) [inline src/ProjectSystem.sln]: can ignore this change

**Comment by mavasani** [comment]
@davkean @basoundr - Addressed your feedback.

**Comment by basoundr** [comment]
:+1:

**Comment by mavasani** [comment]
Any further feedback?

**Comment by davkean** [comment]
:+1:
